goprocess/gp.go

```diff
@@ -33,39 +33,50 @@ func FindAll() []P {
 		return nil
 	}
 
+	found := make(chan P)
+	psCh := make(chan ps.Process)
+	go func() {
+		for _, pr := range pss {
+			psCh <- pr
+		}
+		close(psCh)
+	}()
+
 	var wg sync.WaitGroup
 	wg.Add(len(pss))
-	found := make(chan P)
-	limitCh := make(chan struct{}, concurrencyProcesses)
 
-	for _, pr := range pss {
-		limitCh <- struct{}{}
-		pr := pr
+	for i := 0; i < concurrencyProcesses; i++ {
```

Comment:
Please surround the whole for-loop with a go-routine instead of spawning new go-routines inside. The whole idea of doing that is to ensure that you start reading from `found` without being blocked on the same go-routine.

Reply:
Of course your solution does exactly that, but it spawns unnecessary go-routines.

Reply:
Btw, just for fun, one option, which would allow to spawn less go routines, and also avoid buffered channels. As I personally don't like buffered channels at all -- using them imho usually means that there is something wrong with the design.

```golang
var wg sync.WaitGroup
in := make(chan ps.Process)
out := make(chan P)
// Create fixed amount of workers.
for i := 0; i < concurrencyProcesses; i++ {
  wg.Add(1)
  go func() {
    defer wg.Done()
    for pr := range input {
      p, ok, err := FindProcess(pr.PID())
      // Handle err and !ok.
      out <- p
    }
  }
}
// Create collector.
result := make(chan []P)
go func() {
  defer wg.Done()
  var results []P
  for p := range out {
    results = append(results, p)
  }
  result <- results
}
// Feed processes. There are only so many workers, so we are good.
for _, pr := pss {
  in <- pr
}
close(input)
wg.Wait() // No more workers (producers). Safe to close the output as well.
close(out)
return <-result
```

Reply:
https://github.com/yarcat/gops/blob/fix-123-no-buffers/goprocess/gp.go

Reply:
Yesterday something was completely wrong with me reading messages and code of other people. I am sorry for that. I've opened this change today, and realized that it does *exactly* the same as the one that I wrote here. I am sorry for that. Please ignore all of these comments.